File: browser_extensions/edge_extension/chrome_extension/src/services/master/master_wallet_service.js
/**
 * Master Wallet Service - Browser Extension (Chrome)
 * 
 * Complete Master Wallet Features:
 * - Master wallet creation and ownership
 * - User wallet ownership (master owns all user wallets)
 * - HD Wallet (BIP-39/32/44)
 * - Multi-blockchain support
 * - Token management
 * - Transaction approval
 * - Profit sharing (20% to Super Admin)
 * 
 * This service MUST be identical across ALL platforms.
 */

class MasterWalletService {
  static instance = null;

  static getInstance() {
    if (!MasterWalletService.instance) {
      MasterWalletService.instance = new MasterWalletService();
      MasterWalletService.instance.initialize();
    }
    return MasterWalletService.instance;
  }

  constructor() {
    this.masterWallets = new Map();
    this.userWallets = new Map();
    this.networks = [];
    this.tokens = [];
    this.profitRecords = [];
    
    // Super Admin wallet - MANDATORY 20% profit share
    this.SUPERADMIN_WALLET = "0x742d35Cc6634C0532925a3b844Bc9e7595f1234";
    this.PROFIT_SHARE_PERCENT = 20;
  }

  initialize() {
    this.loadFromStorage();
    this.initializeDefaultNetworks();
  }

  // Load from chrome storage
  async loadFromStorage() {
    try {
      const result = await chrome.storage.local.get([
        'master_wallets',
        'user_wallets',
        'networks',
        'tokens',
        'profit_records'
      ]);
      
      if (result.master_wallets) {
        result.master_wallets.forEach(w => this.masterWallets.set(w.id, w));
      }
      if (result.user_wallets) {
        result.user_wallets.forEach(w => this.userWallets.set(w.id, w));
      }
      if (result.networks) this.networks = result.networks;
      if (result.tokens) this.tokens = result.tokens;
      if (result.profit_records) this.profitRecords = result.profit_records;
    } catch (e) {
      console.error('Failed to load from storage:', e);
    }
  }

  // Save to chrome storage
  async saveToStorage() {
    try {
      await chrome.storage.local.set({
        master_wallets: Array.from(this.masterWallets.values()),
        user_wallets: Array.from(this.userWallets.values()),
        networks: this.networks,
        tokens: this.tokens,
        profit_records: this.profitRecords
      });
    } catch (e) {
      console.error('Failed to save to storage:', e);
    }
  }

  initializeDefaultNetworks() {
    if (this.networks.length === 0) {
      this.networks = [
        { id: 'ethereum', name: 'Ethereum', symbol: 'ETH', chainId: 1, rpcUrl: 'https://eth.llamarpc.com', isEVM: true },
        { id: 'polygon', name: 'Polygon', symbol: 'MATIC', chainId: 137, rpcUrl: 'https://polygon-rpc.com', isEVM: true },
        { id: 'bsc', name: 'BNB Chain', symbol: 'BNB', chainId: 56, rpcUrl: 'https://bsc-dataseed.binance.org', isEVM: true },
        { id: 'arbitrum', name: 'Arbitrum', symbol: 'ETH', chainId: 42161, rpcUrl: 'https://arb1.arbitrum.io/rpc', isEVM: true },
        { id: 'optimism', name: 'Optimism', symbol: 'ETH', chainId: 10, rpcUrl: 'https://mainnet.optimism.io', isEVM: true },
        { id: 'avalanche', name: 'Avalanche', symbol: 'AVAX', chainId: 43114, rpcUrl: 'https://api.avax.network/ext/bc/C/rpc', isEVM: true },
        { id: 'base', name: 'Base', symbol: 'ETH', chainId: 8453, rpcUrl: 'https://mainnet.base.org', isEVM: true },
        { id: 'solana', name: 'Solana', symbol: 'SOL', chainId: 0, rpcUrl: 'https://api.mainnet-beta.solana.com', isEVM: false },
        { id: 'tron', name: 'Tron', symbol: 'TRX', chainId: 0, rpcUrl: 'https://api.trongrid.io', isEVM: false },
        { id: 'bitcoin', name: 'Bitcoin', symbol: 'BTC', chainId: 0, rpcUrl: 'https://blockstream.info/api', isEVM: false }
      ];
    }
  }

  // ============================================================================
  // MASTER WALLET - The Owner
  // ============================================================================

  createMasterWallet(name, type, blockchain) {
    const wallet = {
      id: this.generateId(),
      name: name,
      type: type, // 'hot', 'cold', 'operations'
      blockchain: blockchain,
      address: this.generateAddress(blockchain),
      publicKey: this.generatePublicKey(),
      balance: 0,
      superadminShareAddress: this.SUPERADMIN_WALLET,
      profitSharePercent: this.PROFIT_SHARE_PERCENT,
      isActive: true,
      createdAt: Date.now()
    };
    
    this.masterWallets.set(wallet.id, wallet);
    this.saveToStorage();
    return wallet;
  }

  getMasterWallets() {
    return Array.from(this.masterWallets.values());
  }

  getMasterWallet(id) {
    return this.masterWallets.get(id);
  }

  updateMasterWallet(id, updates) {
    const wallet = this.masterWallets.get(id);
    if (wallet) {
      Object.assign(wallet, updates);
      this.saveToStorage();
      return wallet;
    }
    return null;
  }

  deleteMasterWallet(id) {
    this.masterWallets.delete(id);
    this.saveToStorage();
  }

  // ============================================================================
  // USER WALLETS - Owned by Master Wallet
  // ============================================================================

  createUserWallet(masterWalletId, userId, blockchain) {
    const masterWallet = this.masterWallets.get(masterWalletId);
    if (!masterWallet) {
      throw new Error('Master wallet not found');
    }

    const userWallet = {
      id: this.generateId(),
      userId: userId,
      masterWalletId: masterWalletId,
      ownerMasterWalletAddress: masterWallet.address, // Master OWNS this
      blockchain: blockchain,
      address: this.generateAddress(blockchain),
      publicKey: this.generatePublicKey(),
      balance: 0,
      isActive: true,
      createdAt: Date.now()
    };

    this.userWallets.set(userWallet.id, userWallet);
    this.saveToStorage();
    return userWallet;
  }

  getUserWallets(masterWalletId) {
    return Array.from(this.userWallets.values())
      .filter(w => w.masterWalletId === masterWalletId);
  }

  getUserWalletsByUser(userId) {
    return Array.from(this.userWallets.values())
      .filter(w => w.userId === userId);
  }

  getUserWallet(id) {
    return this.userWallets.get(id);
  }

  // MASTER WALLET controls user wallet
  controlUserWallet(masterWalletId, userWalletId) {
    const wallet = this.userWallets.get(userWalletId);
    if (wallet && wallet.masterWalletId === masterWalletId) {
      return wallet;
    }
    return null;
  }

  // MASTER WALLET approves transactions
  approveTransaction(masterWalletId, userWalletId, txHash) {
    const wallet = this.controlUserWallet(masterWalletId, userWalletId);
    return wallet !== null;
  }

  // ============================================================================
  // NETWORKS & TOKENS
  // ============================================================================

  getNetworks() {
    return this.networks;
  }

  addNetwork(network) {
    this.networks.push(network);
    this.saveToStorage();
  }

  getTokens() {
    return this.tokens;
  }

  async loadTokensFromAPI() {
    try {
      const response = await fetch('https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&order=market_cap_desc&per_page=500&page=1&sparkline=false');
      const data = await response.json();
      this.tokens = data.map(coin => ({
        id: coin.id,
        symbol: coin.symbol.toUpperCase(),
        name: coin.name,
        image: coin.image,
        currentPrice: coin.current_price,
        marketCap: coin.market_cap,
        rank: coin.market_cap_rank
      }));
      this.saveToStorage();
    } catch (e) {
      console.error('Failed to load tokens:', e);
    }
  }

  // ============================================================================
  // PROFIT SHARING - 20% to Super Admin
  // ============================================================================

  calculateProfitShare(totalProfit) {
    return {
      totalProfit: totalProfit,
      superadminShare: (totalProfit * this.PROFIT_SHARE_PERCENT) / 100,
      brandingShare: (totalProfit * (100 - this.PROFIT_SHARE_PERCENT)) / 100,
      superadminAddress: this.SUPERADMIN_WALLET
    };
  }

  recordProfit(brandingId, masterWalletId, totalProfit) {
    const share = this.calculateProfitShare(totalProfit);
    
    const record = {
      id: this.generateId(),
      brandingId: brandingId,
      masterWalletId: masterWalletId,
      totalProfit: share.totalProfit,
      superadminShare: share.superadminShare,
      brandingShare: share.brandingShare,
      superadminAddress: this.SUPERADMIN_WALLET,
      timestamp: Date.now()
    };
    
    this.profitRecords.push(record);
    this.saveToStorage();
    return record;
  }

  getProfitRecords(brandingId) {
    return this.profitRecords.filter(r => r.brandingId === brandingId);
  }

  getSuperAdminWallet() {
    return this.SUPERADMIN_WALLET;
  }

  // ============================================================================
  // HELPERS
  // ============================================================================

  generateId() {
    return '0x' + Array(64).fill(0).map(() => 
      Math.floor(Math.random() * 16).toString(16)
    ).join('');
  }

  generateAddress(blockchain) {
    // In production, use proper key derivation
    const prefix = blockchain === 'bitcoin' ? 'bc1' : '0x';
    const chars = '0123456789abcdef';
    let address = prefix;
    for (let i = 0; i < (blockchain === 'bitcoin' ? 38 : 40); i++) {
      address += chars[Math.floor(Math.random() * chars.length)];
    }
    return address;
  }

  generatePublicKey() {
    // In production, derive from HD wallet
    return this.generateAddress('ethereum');
  }
}

// Export for use
if (typeof module !== 'undefined' && module.exports) {
  module.exports = MasterWalletService;
}

File: mobile_apps/android_app/TigerWallet/app/src/main/java/com/tigerwallet/app/master/MasterWalletService.kt
/**
 * TigerWallet Android - Master Wallet Service
 * Master wallet OWNS and CONTROLS all user wallets
 *
 * Fail-closed: wallet `address` and `publicKey` are NEVER fabricated. Wallet
 * creation delegates to the REAL backend (go/wallet_api at
 * BACKEND_BASE_URL, POST /api/v1/wallets), which derives a real secp256k1
 * address from a real BIP-39 mnemonic. The previous implementation returned
 * `"0x" + UUID.randomUUID().replace("-","").take(40)` as the address and
 * `take(130)` as the public key — that fabrication is removed. If the backend
 * is unreachable or rejects the request, or no JWT is configured, creation
 * throws fail-closed (no wallet with a fake address/key is ever persisted).
 */

package com.tigerwallet.app.master

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * Master Wallet Service - Owner of All User Wallets
 */
class MasterWalletService private constructor() {

    companion object {
        val instance: MasterWalletService by lazy { MasterWalletService() }

        /** Real backend base URL (go/wallet_api). Wallet creation is delegated
         *  here so the returned address is a real secp256k1 address, never a
         *  fabricated `"0x" + UUID`. */
        const val BACKEND_BASE_URL = "http://localhost:8443"

        private val JSON_MEDIA_TYPE = "application/json".toMediaType()
    }

    /** JWT used to authenticate backend wallet-creation. Must be set by the
     *  host app before createMasterWallet/createUserWallet can create a wallet.
     *  When empty, creation fails closed. */
    @Volatile
    var jwt: String = ""

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS).build()

    // Master wallet is the OWNER of all user wallets
    private val _masterWalletsFlow = MutableStateFlow<List<MasterWallet>>(emptyList())
    val masterWalletsFlow: StateFlow<List<MasterWallet>> = _masterWalletsFlow.asStateFlow()

    // All user wallets owned by master wallet
    private val _userWalletsFlow = MutableStateFlow<List<UserWallet>>(emptyList())
    val userWalletsFlow: StateFlow<List<UserWallet>> = _userWalletsFlow.asStateFlow()

    private val _networksFlow = MutableStateFlow<List<BlockchainNetwork>>(emptyList())
    val networksFlow: StateFlow<List<BlockchainNetwork>> = _networksFlow.asStateFlow()

    private val _tokensFlow = MutableStateFlow<List<CryptoToken>>(emptyList())
    val tokensFlow: StateFlow<List<CryptoToken>> = _tokensFlow.asStateFlow()

    fun initialize() {
        loadMasterWallets()
        loadUserWallets()
        loadNetworks()
        loadTokens()
    }

    // ============================================================================
    // MASTER WALLET - The Owner
    // ============================================================================

    private fun loadMasterWallets() {
        val data = KeychainManager.load("master_wallets")
        if (data != null) {
            try {
                val wallets = SerializationUtils.deserialize<List<MasterWallet>>(data)
                _masterWalletsFlow.value = wallets
            } catch (e: Exception) { _masterWalletsFlow.value = emptyList() }
        }
    }

    /**
     * Create a master wallet. The address is derived by the REAL backend
     * (POST /api/v1/wallets, real secp256k1 from a real BIP-39 mnemonic) and is
     * never fabricated. `publicKey` is left empty (the backend wallet record
     * exposes only the address; a full public key must come from a real
     * secp256k1 derivation path, never a UUID). Throws fail-closed if the
     * backend is unreachable/rejects the request or no JWT is configured.
     */
    suspend fun createMasterWallet(name: String, type: WalletType, blockchain: String): MasterWallet =
        withContext(Dispatchers.IO) {
            val backend = createWalletViaBackend(name, blockchain)
            val wallet = MasterWallet(
                id = backend.id,
                name = name,
                type = type,
                blockchain = blockchain,
                address = backend.address,
                publicKey = "",
                balance = 0.0,
                isActive = true,
                createdAt = System.currentTimeMillis()
            )
            saveMasterWallet(wallet)
            _masterWalletsFlow.value = _masterWalletsFlow.value + wallet
            wallet
        }

    private fun saveMasterWallet(wallet: MasterWallet) {
        val wallets = _masterWalletsFlow.value.toMutableList()
        wallets.removeAll { it.id == wallet.id }
        wallets.add(wallet)
        val data = SerializationUtils.serialize(wallets)
        KeychainManager.save("master_wallets", data)
    }

    fun getMasterWallets(): List<MasterWallet> = _masterWalletsFlow.value

    // ============================================================================
    // USER WALLETS - Owned by Master Wallet
    // ============================================================================

    private fun loadUserWallets() {
        val data = KeychainManager.load("user_wallets")
        if (data != null) {
            try {
                val wallets = SerializationUtils.deserialize<List<UserWallet>>(data)
                _userWalletsFlow.value = wallets
            } catch (e: Exception) { _userWalletsFlow.value = emptyList() }
        }
    }

    // MASTER WALLET creates/owns user wallets
    suspend fun createUserWallet(masterWalletId: String, userId: String, blockchain: String): UserWallet =
        withContext(Dispatchers.IO) {
            val masterWallet = _masterWalletsFlow.value.find { it.id == masterWalletId }
                ?: throw Exception("Master wallet not found")

            val backend = createWalletViaBackend(masterWallet.name, blockchain)
            val userWallet = UserWallet(
                id = backend.id,
                userId = userId,
                ownerMasterWalletId = masterWalletId,  // OWNERSHIP
                ownerAddress = masterWallet.address,     // Master wallet address owns this
                blockchain = blockchain,
                address = backend.address,
                publicKey = "",
                balance = 0.0,
                isActive = true,
                createdAt = System.currentTimeMillis()
            )

            saveUserWallet(userWallet)
            _userWalletsFlow.value = _userWalletsFlow.value + userWallet
            userWallet
        }

    // MASTER WALLET can control any user wallet
    fun controlUserWallet(masterWalletId: String, userWalletId: String): UserWallet? {
        return _userWalletsFlow.value.find { 
            it.id == userWalletId && it.ownerMasterWalletId == masterWalletId 
        }
    }

    // MASTER WALLET approves all transactions from user wallets
    fun approveTransaction(masterWalletId: String, userWalletId: String, txHash: String): Boolean {
        val userWallet = controlUserWallet(masterWalletId, userWalletId)
        return userWallet != null  // Master wallet approves
    }

    private fun saveUserWallet(wallet: UserWallet) {
        val wallets = _userWalletsFlow.value.toMutableList()
        wallets.removeAll { it.id == wallet.id }
        wallets.add(wallet)
        val data = SerializationUtils.serialize(wallets)
        KeychainManager.save("user_wallets", data)
        _userWalletsFlow.value = wallets
    }

    // Get all user wallets owned by a master wallet
    fun getUserWallets(masterWalletId: String): List<UserWallet> {
        return _userWalletsFlow.value.filter { it.ownerMasterWalletId == masterWalletId }
    }

    // Get all user wallets for a user
    fun getUserWalletsByUser(userId: String): List<UserWallet> {
        return _userWalletsFlow.value.filter { it.userId == userId }
    }

    // ============================================================================
    // Networks & Tokens (same as before)
    // ============================================================================

    private fun loadNetworks() {
        _networksFlow.value = getDefaultNetworks()
    }

    private fun loadTokens() {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val tokens = fetchTokensFromAPI()
                if (tokens.isNotEmpty()) _tokensFlow.value = tokens
            } catch (e: Exception) { _tokensFlow.value = emptyList() }
        }
    }

    private suspend fun fetchTokensFromAPI(): List<CryptoToken> = withContext(Dispatchers.IO) {
        try {
            val request = Request.Builder()
                .url("https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&order=market_cap_desc&per_page=500&page=1&sparkline=false")
                .get().build()
            val response = client.newCall(request).execute()
            if (response.isSuccessful) {
                val body = response.body?.string() ?: return@withContext emptyList()
                val jsonArray = JSONArray(body)
                val tokens = mutableListOf<CryptoToken>()
                for (i in 0 until jsonArray.length()) {
                    val coin = jsonArray.getJSONObject(i)
                    tokens.add(CryptoToken(
                        id = coin.getString("id"),
                        symbol = coin.getString("symbol").uppercase(),
                        name = coin.getString("name"),
                        image = coin.optString("image", ""),
                        currentPrice = coin.optDouble("current_price", 0.0),
                        marketCap = coin.optLong("market_cap", 0L),
                        rank = coin.optInt("market_cap_rank", 0),
                        priceChange24h = coin.optDouble("price_change_24h", 0.0)
                    ))
                }
                tokens
            } else emptyList()
        } catch (e: Exception) { emptyList() }
    }

    fun getDefaultNetworks(): List<BlockchainNetwork> = listOf(
        BlockchainNetwork("ethereum", "Ethereum", "ETH", 1, "https://eth.llamarpc.com", true),
        BlockchainNetwork("polygon", "Polygon", "MATIC", 137, "https://polygon-rpc.com", true),
        BlockchainNetwork("bsc", "BNB Chain", "BNB", 56, "https://bsc-dataseed.binance.org", true),
        BlockchainNetwork("arbitrum", "Arbitrum", "ETH", 42161, "https://arb1.arbitrum.io/rpc", true),
        BlockchainNetwork("optimism", "Optimism", "ETH", 10, "https://mainnet.optimism.io", true),
        BlockchainNetwork("avalanche", "Avalanche", "AVAX", 43114, "https://api.avax.network/ext/bc/C/rpc", true),
        BlockchainNetwork("base", "Base", "ETH", 8453, "https://mainnet.base.org", true),
        BlockchainNetwork("solana", "Solana", "SOL", 0, "https://api.mainnet-beta.solana.com", false),
        BlockchainNetwork("tron", "Tron", "TRX", 0, "https://api.trongrid.io", false),
        BlockchainNetwork("bitcoin", "Bitcoin", "BTC", 0, "https://blockstream.info/api", false)
    )

    fun addNetwork(network: BlockchainNetwork) {
        val updated = _networksFlow.value.toMutableList()
        if (updated.none { it.id == network.id }) { updated.add(network); _networksFlow.value = updated }
    }

    fun addToken(token: CryptoToken) {
        val updated = _tokensFlow.value.toMutableList()
        if (updated.none { it.id == token.id }) { updated.add(token); _tokensFlow.value = updated }
    }

    fun getNetworks(): List<BlockchainNetwork> = _networksFlow.value
    fun getTokens(): List<CryptoToken> = _tokensFlow.value

    /**
     * Password used to encrypt the backend wallet seed (backend requires
     * `password` min 8 chars). Must be set by the host app from user input
     * before createMasterWallet/createUserWallet can create a wallet. When
     * empty, creation fails closed.
     */
    @Volatile
    var creationPassword: String = ""

    /** Result of a real backend wallet creation. `address` is a real
     *  secp256k1 address derived by go/wallet_api from a real BIP-39 mnemonic. */
    private data class BackendWallet(val id: String, val address: String)

    private fun chainIdFor(blockchain: String): Long =
        getDefaultNetworks().firstOrNull { it.id == blockchain }?.chainId
            ?: getDefaultNetworks().firstOrNull { it.id == "ethereum" }?.chainId
            ?: 1L

    /**
     * Creates a wallet on the REAL backend (POST /api/v1/wallets) and returns
     * the real secp256k1 address + wallet id. Never fabricates an address or
     * public key. Throws fail-closed if no JWT/password is configured, the
     * backend is unreachable, or it rejects the request.
     */
    private fun createWalletViaBackend(label: String, blockchain: String): BackendWallet {
        if (jwt.isEmpty()) {
            throw IllegalStateException(
                "No auth JWT configured; cannot create wallet on backend (fail-closed)."
            )
        }
        if (creationPassword.length < 8) {
            throw IllegalStateException(
                "Wallet creation requires a user password (>=8 chars) for backend seed encryption (fail-closed)."
            )
        }
        val body = JSONObject().apply {
            put("label", label)
            put("chain_id", chainIdFor(blockchain))
            put("password", creationPassword)
            put("entropy_bits", 256)
        }
        val request = Request.Builder()
            .url("$BACKEND_BASE_URL/api/v1/wallets")
            .header("Authorization", "Bearer $jwt")
            .header("Content-Type", "application/json")
            .post(body.toString().toRequestBody(JSON_MEDIA_TYPE))
            .build()
        val respData: String
        val statusCode: Int
        try {
            client.newCall(request).execute().use { resp ->
                statusCode = resp.code
                respData = resp.body?.string() ?: ""
            }
        } catch (e: Exception) {
            throw IllegalStateException("Backend unreachable: ${e.message ?: e.toString()}")
        }
        if (statusCode != 201) {
            throw IllegalStateException("Backend rejected wallet creation (HTTP $statusCode): $respData")
        }
        val json: JSONObject
        try {
            json = JSONObject(respData)
        } catch (e: Exception) {
            throw IllegalStateException("Malformed backend response: $respData")
        }
        val id = json.optString("id", "")
        val address = json.optString("address", "")
        if (id.isEmpty() || !address.startsWith("0x") || address.length != 42) {
            throw IllegalStateException("Backend did not return a valid wallet address: $respData")
        }
        return BackendWallet(id, address)
    }
}

// ============================================================================
// Data Models
// ============================================================================

enum class WalletType { HOT, COLD, OPERATIONS }

data class MasterWallet(
    val id: String,
    val name: String,
    val type: WalletType,
    val blockchain: String,
    val address: String,
    val publicKey: String,
    val balance: Double,
    val isActive: Boolean,
    val createdAt: Long
)

// User wallet OWNED by master wallet
data class UserWallet(
    val id: String,
    val userId: String,
    val ownerMasterWalletId: String,   // WHO OWNS THIS WALLET
    val ownerAddress: String,          // Master wallet address
    val blockchain: String,
    val address: String,
    val publicKey: String,
    val balance: Double,
    val isActive: Boolean,
    val createdAt: Long
)

data class BlockchainNetwork(val id: String, val name: String, val symbol: String, val chainId: Long, val rpcUrl: String, val isEVM: Boolean)
data class CryptoToken(val id: String, val symbol: String, val name: String, val image: String, val currentPrice: Double, val marketCap: Long, val rank: Int, val priceChange24h: Double)

// Utilities
object KeychainManager {
    private val prefs = android.content.Context.getSharedPreferences("tigerwallet", android.content.Context.MODE_PRIVATE)
    fun save(key: String, data: ByteArray) { prefs.edit().putString(key, android.util.Base64.encodeToString(data, android.util.Base64.NO_WRAP)).apply() }
    fun load(key: String): ByteArray? { val encoded = prefs.getString(key, null) ?: return null; return android.util.Base64.decode(encoded, android.util.Base64.NO_WRAP) }
}

object SerializationUtils {
    private val gson = com.google.gson.Gson()
    fun serialize(obj: Any): ByteArray = gson.toJson(obj).toByteArray()
    inline fun <reified T> deserialize(data: ByteArray): T = gson.fromJson(String(data), T::class.java)
}
